### PR TITLE
Get events

### PR DESCRIPTION
**Pull request by marioiordanov**
## Usage related changes

starknet_getEvents rpc endpoint

## Development related changes

- Updated starknet_in_rust version
- Added method for extracting blocks based on from/to BlockId criteria
- Refactored block to hold TransactionHash, instead of Transaction
- Added utility method for integration testing that makes a get request to devnet
- Added logging of rpc method name, when transaction is received
- Added method for extracting events from transaction
- StarknetState passes default values instead of None, due to bug in starknet_in_rust about fee estimation

## Checklist:

- [x] Applied formatting - `./scripts/format.sh`
- [x] No linter errors - `./scripts/clippy_check.sh`
- [x] Performed code self-review
- [x] Rebased to the last commit of the target branch (or merged it into my branch)
- [x] Documented the changes
- [x] Linked the issues which this PR resolves
- [x] Updated the tests
- [x] All tests are passing - `cargo test`

### CODE REVIEW

File: crates/starknet/src/blocks/mod.rs [hunk 1]
```diff
@@ -164,13 +210,329 @@ impl HashProducer for StarknetBlock {
...
+                ))
+                .is_some()
+        );
+        assert!(blocks.block_number_from_block_id(BlockId::Hash(block_hash.into())).is_some());
+    }
+
+    #[test]
+    fn get_blocks_with_filter() {
+        let mut blocks = StarknetBlocks::default();
+
```
Comment (mikiw) [inline crates/starknet/src/blocks/mod.rs:269]: asserting length is not enough, we would need to assert real data like hashes or something

Comment (marioiordanov) [reply crates/starknet/src/blocks/mod.rs:315]: After this comment many changes were pushed. Can you rereview it

File: crates/starknet/src/starknet/mod.rs [hunk 2]
```diff
@@ -521,6 +523,25 @@ impl Starknet {
 
         Ok(block.clone())
     }
+
+    pub fn get_transaction_by_hash(&self, transaction_hash: Felt) -> Result<&Transaction> {
+        self.transactions
+            .get_by_hash(transaction_hash)
+            .map(|starknet_transaction| &starknet_transaction.inner)
+            .ok_or(crate::error::Error::NoTransaction)
```
Comment (taco-paco) [inline crates/starknet/src/starknet/mod.rs:531]:
```suggestion
            .ok_or(Error::NoTransaction)
```

File: crates/starknet/src/starknet/events.rs [hunk 3]
```diff
@@ -0,0 +1,438 @@
...
+    to_block: Option<BlockId>,
+    contract_address: Option<ContractAddress>,
+    keys_filter: Option<Vec<Vec<Felt>>>,
+    mut skip: usize,
+    limit: Option<usize>,
+) -> Result<(Vec<EmittedEvent>, bool)> {
+    let blocks = starknet.blocks.get_blocks(from_block, to_block)?;
+    let mut events: Vec<EmittedEvent> = Vec::new();
+    // convert to starknet_in_rust::utils::Address
+    let address =
```
[the remark this answers is not in the document]
Comment (marioiordanov) [reply crates/starknet/src/starknet/events.rs:45]: if From trait was implemented, it would have worked

[the remark this answers is not in the document]
Comment (marioiordanov) [reply crates/starknet/src/starknet/events.rs:45]: it doenst compile this way

File: crates/starknet/src/starknet/events.rs [hunk 4]
```diff
@@ -0,0 +1,438 @@
...
+    let mut elements_added = 0;
+
+    // iterate over each block and get the transactions for each one
+    // then iterate over each transaction events and filter them
+    for block in blocks {
+        for transaction_hash in block.get_transactions() {
+            let transaction = starknet
+                .transactions
+                .get_by_hash(*transaction_hash)
+                .ok_or(crate::error::Error::NoTransaction)?;
```
[the remark this answers is not in the document]
Comment (marioiordanov) [reply crates/starknet/src/starknet/events.rs]: [87bcd4d](https://github.com/0xSpaceShard/starknet-devnet-rs/pull/117/commits/87bcd4d4722e324b079057316c7a7142380d017b)

File: crates/starknet/src/starknet/mod.rs (hunk 2 above)
Comment (marioiordanov) [reply crates/starknet/src/starknet/mod.rs:531]: [87bcd4d](https://github.com/0xSpaceShard/starknet-devnet-rs/pull/117/commits/87bcd4d4722e324b079057316c7a7142380d017b)

File: crates/starknet/src/starknet/events.rs [hunk 5]
```diff
@@ -0,0 +1,438 @@
+use starknet_api::block::BlockNumber;
+use starknet_in_rust::execution::Event;
+use starknet_in_rust::felt::Felt252;
+use starknet_in_rust::utils::Address;
+use starknet_rs_core::types::BlockId;
+use starknet_types::contract_address::ContractAddress;
+use starknet_types::felt::Felt;
+
+use super::Starknet;
+use crate::error::{self, Result};
```
[the remark this answers is not in the document]
Comment (marioiordanov) [reply crates/starknet/src/starknet/events.rs]: [87bcd4d](https://github.com/0xSpaceShard/starknet-devnet-rs/pull/117/commits/87bcd4d4722e324b079057316c7a7142380d017b)

File: crates/starknet/src/starknet/events.rs [hunk 6]
```diff
@@ -0,0 +1,438 @@
...
+                .ok_or(crate::error::Error::NoTransaction)?;
+
+            // filter the events from the transaction
+            let filtered_transaction_events = transaction
+                .get_events()?
+                .into_iter()
+                .filter(|event| check_if_filter_applies_for_event(&address, &keys_filter, event));
+
+            // produce an emitted event for each filtered transaction event
+            for transaction_event in filtered_transaction_events.skip_while(|_| {
```
[the remark this answers is not in the document]
Comment (marioiordanov) [reply crates/starknet/src/starknet/events.rs]: [8b3adcb](https://github.com/0xSpaceShard/starknet-devnet-rs/pull/117/commits/8b3adcb83b1cbcc7ed8bdbb9ed2ffd2d3dec1589)

File: crates/starknet/src/starknet/mod.rs (hunk 2 above)
Comment (marioiordanov) [reply crates/starknet/src/starknet/mod.rs]: [50b0196](https://github.com/0xSpaceShard/starknet-devnet-rs/pull/117/commits/50b0196f41e94966a3ecde8d76d4ce7b78224e2d)

File: crates/starknet-server/src/api/json_rpc/endpoints.rs [hunk 7]
```diff
@@ -324,8 +321,42 @@ impl JsonRpcHandler {
...
 
     /// starknet_getEvents
-    pub(crate) async fn get_events(&self, _filter: EventFilter) -> RpcResult<EventsChunk> {
-        Err(error::ApiError::InvalidContinuationToken)
+    pub(crate) async fn get_events(&self, filter: EventFilter) -> RpcResult<EventsChunk> {
+        let starknet = self.api.starknet.read().await;
+
+        let page = filter
+            .continuation_token
+            .unwrap_or("0".to_string())
```
[the remark this answers is not in the document]
Comment (marioiordanov) [reply crates/starknet-server/src/api/json_rpc/endpoints.rs]: [bf52d71](https://github.com/0xSpaceShard/starknet-devnet-rs/pull/117/commits/bf52d71725849f8f13951525ab16d975c359be71)

File: crates/starknet-server/src/api/json_rpc/endpoints.rs [hunk 8]
```diff
@@ -324,8 +321,42 @@ impl JsonRpcHandler {
     }
 
     /// starknet_getEvents
-    pub(crate) async fn get_events(&self, _filter: EventFilter) -> RpcResult<EventsChunk> {
-        Err(error::ApiError::InvalidContinuationToken)
+    pub(crate) async fn get_events(&self, filter: EventFilter) -> RpcResult<EventsChunk> {
+        let starknet = self.api.starknet.read().await;
+
+        let page = filter
```
[the remark this answers is not in the document]
Comment (marioiordanov) [reply crates/starknet-server/src/api/json_rpc/endpoints.rs:329]: which one

File: crates/starknet-server/src/api/models/transaction.rs [hunk 9]
```diff
@@ -232,19 +230,27 @@ pub struct EventContent {
...
 
 #[derive(Debug, Clone, PartialEq, Eq, Deserialize, Serialize)]
 pub struct EventsChunk {
-    pub events: Vec<Event>,
+    pub events: Vec<EmittedEvent>,
     pub continuation_token: Option<String>,
 }
 
+#[derive(Debug, Clone, PartialEq, Eq, Deserialize, Serialize)]
+pub struct EmittedEvent {
```
Comment (taco-paco) [inline crates/starknet-server/src/api/models/transaction.rs:246]: It would be good to define new types in `types` crate. To save effort of refactoring

Comment (marioiordanov) [reply crates/starknet-server/src/api/models/transaction.rs:246]: I will add it

Comment (marioiordanov) [reply crates/starknet-server/src/api/models/transaction.rs:246]: https://github.com/0xSpaceShard/starknet-devnet-rs/commit/f54fb47b91f7a0761329c17406a6a4403cd0c7f0

File: crates/starknet/src/starknet/events.rs [hunk 10]
```diff
@@ -0,0 +1,438 @@
...
+use starknet_in_rust::utils::Address;
+use starknet_rs_core::types::BlockId;
+use starknet_types::contract_address::ContractAddress;
+use starknet_types::felt::Felt;
+
+use super::Starknet;
+use crate::error::{self, Result};
+use crate::traits::HashIdentified;
+
+pub struct EmittedEvent {
```
[the remark this answers is not in the document]
Comment (marioiordanov) [reply crates/starknet/src/starknet/events.rs]: https://github.com/0xSpaceShard/starknet-devnet-rs/commit/f54fb47b91f7a0761329c17406a6a4403cd0c7f0

**Review by mikiw — COMMENTED** [review]
I'm not a rust expert but from an old devnet perspective looks legit to me.

File: crates/starknet/src/starknet/events.rs [hunk 11]
```diff
@@ -0,0 +1,425 @@
...
+        let (events, has_more) =
+            get_events(&starknet, None, Some(BlockId::Number(1)), None, None, 3, Some(1)).unwrap();
+        assert_eq!(events.len(), 0);
+        assert!(!has_more);
+
+        // returns all events from the first 2 blocks, skip 1, but limit the result to 1, it should
+        // return 1 event and should be more
+        let (events, has_more) =
+            get_events(&starknet, None, Some(BlockId::Number(1)), None, None, 1, Some(1)).unwrap();
+        assert_eq!(events.len(), 1);
```
[the remark this answers is not in the document]
Comment (marioiordanov) [reply crates/starknet/src/starknet/events.rs:370]: what do you mean? please explain to me

File: crates/starknet/src/starknet/events.rs [hunk 12]
```diff
@@ -0,0 +1,425 @@
...
+    }
+
+    #[test]
+    fn pagination_for_events_of_multiple_blocks() {
+        let starknet = setup();
+
+        // returns all events from all blocks
+        let (events, has_more) = get_events(&starknet, None, None, None, None, 0, None).unwrap();
+
+        assert_eq!(events.len(), 15);
```
Comment (mikiw) [inline crates/starknet/src/starknet/events.rs:351]: In many places in tests, there are assertions with .len(). For example here, we are getting 15 events but actually, we don't check if the events are correct. The wrong implementation of adaptations to the next version can return just a vector with 15 empty record and the test will still pass.

File: crates/starknet/src/blocks/mod.rs (hunk 1 above)
Comment (marioiordanov) [reply crates/starknet/src/blocks/mod.rs:315]: There is already such test https://github.com/0xSpaceShard/starknet-devnet-rs/blob/6c53ed1703b03efd59eeae5ba41b40c86081b95c/crates/starknet/src/blocks/mod.rs#L234

File: crates/starknet/src/starknet/events.rs (hunk 11 above)
[the remark this answers is not in the document]
Comment (mikiw) [reply crates/starknet/src/starknet/events.rs:370]: something like this:
assert_eq!(events[0].event.keys[0], ["0x123"]);

similar to https://github.com/0xSpaceShard/starknet-devnet/blob/44a9a030908f915bb7642c3fbdbc66d014ac078a/test/rpc/test_data/get_events.py

[the remark this answers is not in the document]
Comment (mikiw) [reply crates/starknet/src/starknet/events.rs:370]: here is usage of GET_EVENTS_TEST_DATA:
https://github.com/0xSpaceShard/starknet-devnet/blob/44a9a030908f915bb7642c3fbdbc66d014ac078a/test/rpc/test_rpc_misc.py#L320s

[the remark this answers is not in the document]
Comment (marioiordanov) [reply crates/starknet/src/starknet/events.rs:370]: added test https://github.com/0xSpaceShard/starknet-devnet-rs/blob/84a8d0639718ab4d00f552a4fdabb86c0cd18c6e/crates/starknet/src/starknet/events.rs#L375

File: crates/starknet/src/starknet/events.rs [hunk 13]
```diff
@@ -0,0 +1,425 @@
...
+        assert!(!has_more);
+
+        // returns all events from all blocks, skip 3, but limit the result to 10
+        let (events, has_more) =
```
[the remark this answers is not in the document]
Comment (marioiordanov) [reply crates/starknet/src/starknet/events.rs:355]: https://github.com/0xSpaceShard/starknet-devnet-rs/commit/1960d117b24ef61430e9e9a561b646e6b32e6a00

File: crates/starknet-server/src/api/json_rpc/endpoints.rs [hunk 14]
```diff
@@ -42,28 +42,24 @@ impl JsonRpcHandler {
...
     }
 
     /// starknet_getBlockWithTxs
     pub(crate) async fn get_block_with_txs(&self, block_id: BlockId) -> RpcResult<Block> {
-        let block =
-            self.api.starknet.read().await.get_block(block_id.into()).map_err(|err| match err {
-                Error::NoBlock => ApiError::BlockNotFound,
-                unknown_error => ApiError::StarknetDevnetError(unknown_error),
-            })?;
+        let starknet = self.api.starknet.read().await;
```
[the remark this answers is not in the document]
Comment (marioiordanov) [reply crates/starknet-server/src/api/json_rpc/endpoints.rs:54]: moved in [6f8528f](https://github.com/0xSpaceShard/starknet-devnet-rs/pull/117/commits/6f8528f72d3ac5d01f86e9c99cf5565cbfc591e5)

**Review by taco-paco — APPROVED** [review]

**Pull request merged**